In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
df=pd.read_csv('/content/Churn_Modelling.csv')

In [ ]:
df.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [ ]:
X=df.iloc[:,3:13].values
y=df.iloc[:,13].values

In [ ]:
X

array([[619, 'France', 'Female', ..., 1, 1, 101348.88],
       [608, 'Spain', 'Female', ..., 0, 1, 112542.58],
       [502, 'France', 'Female', ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 'Female', ..., 0, 1, 42085.58],
       [772, 'Germany', 'Male', ..., 1, 0, 92888.52],
       [792, 'France', 'Female', ..., 1, 0, 38190.78]], dtype=object)

In [ ]:
y

array([1, 0, 1, ..., 1, 1, 0])

In [ ]:
#encoding categorical data
from sklearn.preprocessing import LabelEncoder,OneHotEncoder
LabelEncoder_X_1=LabelEncoder()
X[:,1]=LabelEncoder_X_1.fit_transform(X[:,1])
LabelEncoder_X_2=LabelEncoder()
X[:,2]=LabelEncoder_X_2.fit_transform(X[:,2])

In [ ]:
X

array([[619, 0, 0, ..., 1, 1, 101348.88],
       [608, 2, 0, ..., 0, 1, 112542.58],
       [502, 0, 0, ..., 1, 0, 113931.57],
       ...,
       [709, 0, 0, ..., 0, 1, 42085.58],
       [772, 1, 1, ..., 1, 0, 92888.52],
       [792, 0, 0, ..., 1, 0, 38190.78]], dtype=object)

In [ ]:
#splitting the dataset into training and testing
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=0)

In [ ]:
len(X_train)

8000

In [ ]:
len(X_test)

2000

In [ ]:
#feature scalling
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train=sc.fit_transform(X_train)
X_test=sc.transform(X_test)

In [ ]:
#classical ML
from sklearn.tree import DecisionTreeClassifier
clf=DecisionTreeClassifier()
clf.fit(X_train,y_train)
pred=clf.predict(X_test)

In [ ]:
#making confusion matrix
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,pred)
print(cm)

[[1358  237]
 [ 185  220]]


In [ ]:
#accuracy score
from sklearn.metrics import accuracy_score
score=accuracy_score(y_test,pred)
print(score*100)

78.9


#Accuracy from decision Tree is 78.9

In [ ]:
#Random forest
from sklearn.ensemble import RandomForestClassifier
clf2=RandomForestClassifier(n_estimators=200)
clf2.fit(X_train,y_train)
pred2=clf2.predict(X_test)

In [ ]:
#making confusion matrix
from sklearn.metrics import confusion_matrix
cm2=confusion_matrix(y_test,pred2)
print(cm2)

[[1523   72]
 [ 201  204]]


In [ ]:
#accuracy score
from sklearn.metrics import accuracy_score
score2=accuracy_score(y_test,pred2)
print(score2*100)

86.35000000000001


#Accuracy from Random forest is 86.3

In [ ]:
#ANN

In [ ]:
import keras
from keras.models import Sequential
from keras.layers import Dense

In [ ]:
#initializing ANN
clasifier=Sequential()
#units=input+output/2
# initilizer=how ur weights are updated
# relu activation function
# input_dim=input shape(how many feature do u have in the data)

#adding input layer and first hidden layer
clasifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
#adding second hidden layer
clasifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
#adding third hidden layer
clasifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
#adding output layer
clasifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
#compile ANN
clasifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train[0]
#fitting the ANN training to Training set
clasifier.fit(X_train,y_train,batch_size=10,epochs=50)

Epoch 1/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.7959 - loss: 0.4900
Epoch 2/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7960 - loss: 0.4345
Epoch 3/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.4294
Epoch 4/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8145 - loss: 0.4241
Epoch 5/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8278 - loss: 0.4205
Epoch 6/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8300 - loss: 0.4178
Epoch 7/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8301 - loss: 0.4157
Epoch 8/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8316 - loss: 0.4135
Epoch 9/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8322 - loss: 0.4116
Epoch 10/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8317 - loss: 0.4099
Epoch 11/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8328 - loss: 0.4082
Epoch 12/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step

In [ ]:
#predicting the model
y_pred_ann=clasifier.predict(X_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step
[[1534   61]
 [ 255  150]]


In [ ]:
#accuracy
score3=accuracy_score(y_test,y_pred_ann)
print(score3*100)

84.2


#Accuracy for ANN is 84.2

#conclusion
So, as we have already seen, by using 2 hidden layered ANN, we aren't able to get beyond 85% accuracy, ofcourse we can get higher accuracy by increasing the epochs & doing some hyperparaméter tuning, so on the other hand, the classical models can also be tuned further.

Bottomline, always start with classical machine learning models, and if we aren't able to increase the performance, then move towards Deep Learning techniques :)